pyanaconda/modules/subscription/runtime.py

```diff
@@ -0,0 +1,111 @@
+    CONFIG_KEY_SERVER_PROXY_USER = "server.proxy_user"
+    CONFIG_KEY_SERVER_PROXY_PASSWORD = "server.proxy_password"
+    CONFIG_KEY_RHSM_BASEURL = "rhsm.baseurl"
+
+    def __init__(self, rhsm_config_proxy, rhsm_config_defaults, subscription_request):
+        """Create a new task for setting RHSM configuration.
+
+        :param rhsm_config_proxy: DBus proxy for the RHSM Config object
+        :param dict rhsm_config_defaults: a dictionary of original RHSM configuration values
+        :param subscription_request: subscription request DBus
+        :type subscription_request: SubscriptionRequest instance
+        """
+        super().__init__()
+        self._rhsm_config_proxy = rhsm_config_proxy
+        self._request = subscription_request
+        self._rhsm_config_defaults = rhsm_config_defaults
+
+    @property
+    def name(self):
+        return "Set RHSM configuration."
+
+    def run(self):
+        log.debug("subscription: setting RHSM config values")
+        # We will use the SetAll() dbus method and we need to
+        # assemble a dictionary that we will feed to it.
+        # Start by preparing a SubscriptionData property mapping
+        # to the RHSM config keys.
+        #
+        # A note about constructing the dict:
+        # - DBus API needs all values to be strings, so we need to convert the
+        #   port number to string
+        # - proxy password is stored in SecretData instance and we need to retrieve
+        #   its value
+        property_key_map = {
+            self.CONFIG_KEY_SERVER_HOSTNAME: self._request.server_hostname,
+            self.CONFIG_KEY_SERVER_PROXY_HOSTNAME: self._request.server_proxy_hostname,
+            self.CONFIG_KEY_SERVER_PROXY_PORT: str(self._request.server_proxy_port),
+            self.CONFIG_KEY_SERVER_PROXY_USER: self._request.server_proxy_user,
+            self.CONFIG_KEY_SERVER_PROXY_PASSWORD: self._request.server_proxy_password.value,
+            self.CONFIG_KEY_RHSM_BASEURL: self._request.rhsm_baseurl
+        }
+
+        # Then process the mapping into the final dict we will set to RHSM. This includes
+        # checking if some values have been cleared by the user and should be restored to
+        # the original values that have been in the RHSM config before we started
+        # manipulating it.
+        config_dict = {}
+        for key, value in property_key_map.items():
+            if value:
+                # if value is present in request, use it
+                config_dict[key] = value
+            else:
+                # if no value is present in request, use
+                # value from the original RHSM config state
+                # (if any)
+                log.debug("subscription: restoring original value for RHSM config key %s", key)
+                config_dict[key] = self._rhsm_config_defaults.get(key, "")
+
+        # and finally set the dict to RHSM via the DBus API
+        self._rhsm_config_proxy.SetAll(config_dict, "")
```

Comment:
I don't see the DBus method `SetAll` in the documentation of the RHSM Configuration object. Are you sure this is working? I see only the method `Set` and it requires a variant.

Reply:
> don't see the DBus method `SetAll` in the documentation of the RHSM Configuration object. Are you sure this is working? 

Yes, it was recently added but for some reason has not yet shown up in the DBus API documentation:
https://github.com/candlepin/subscription-manager/pull/2224
I have reported this to the Subscription Manager team on our Tuesday sync meeting, so it should be fixed soon.

>I see only the method `Set` and it requires a variant.

So this is the definition for SetAll as far as I can tell:

```python3
    @util.dbus_service_method(
        constants.CONFIG_INTERFACE,
        in_signature='a{sv}s')
    @util.dbus_handle_exceptions
    def SetAll(self, configuration, locale, sender=None):
        """
        Method for setting multiple configuration options. Of course all of them could be set.
        :param configuration: d-bus dictionary with configuration. Keys have to include section. e.g.
                              server.hostname. Configuration file is saved to the file at the end of method.
        :param locale: string with locale
        :param sender: not used
        :return: None
        """
        configuration = dbus_utils.dbus_to_python(configuration, expected_type=dict)
        locale = dbus_utils.dbus_to_python(locale, expected_type=str)
        Locale.set(locale)

```
So `a{sv}s` = dictionary with string to variant mapping (for the config values) & string (for locale). So is it enough to pass it a dictionary (which has string keys and only string values) and dasbus will do the magic, or do I need to do some more processing on it ?

Reply:
Ah, ok. But you still have to create the variants. Also `Get` and `GetAll` return variants, but you can just use `get_native` for the return values.

Dasbus doesn't unpack the variants, so the arguments and return values match the signatures of the methods. Pydbus used to unpack the received variants, but still required to provide variants to send. It was a mess.

Reply:
OK, thanks for deciphering it for me. :) So I've added `get_native()` over the `GetAll()` call and changed the `SetAll()` invocation to use a `Dict[Str, Str]` variant. I hope that's correct ?